geode-core/src/upgradeTest/java/org/apache/geode/security/AuthExpirationFunctionDUnitTest.java

```diff
@@ -65,77 +76,327 @@
-      ResultCollector rc = onServer(clientCache.getDefaultPool()).execute(writeFunction);
-      assertThat(((ArrayList) rc.getResult()).get(0))
-          .isEqualTo(TestFunctions.WriteFunction.SUCCESS_OUTPUT);
-    });
+    UpdatableUserAuthInitialize.setUser("data2");
+    rc = onServers(clientCache.getDefaultPool()).execute(writeFunction);
+    assertThat(((ArrayList) rc.getResult()).get(0))
+        .isEqualTo(TestFunctions.WriteFunction.SUCCESS_OUTPUT);
+
+    // all function invocation authorizations are recorded
+    List<Object> resultsVM0 = collectSecurityManagerResults(serverVM0);
+    List<Object> resultsVM1 = collectSecurityManagerResults(serverVM1);
+    List<Object> resultsVM2 = collectSecurityManagerResults(serverVM2);
+
+    Set<String> combinedExpiredUsers = combineExpiredUsers(resultsVM0, resultsVM1, resultsVM2);
+
+    assertThat(combinedExpiredUsers.size()).isEqualTo(1);
+    assertThat(combinedExpiredUsers.contains("data1")).isTrue();
+    Map<String, List<String>> authorizedOps = collectVMOps(1, resultsVM0, resultsVM1, resultsVM2);
+
+    assertThat(authorizedOps.get("data1")).asList().hasSize(3);
+    assertThat(authorizedOps.get("data1")).asList().containsExactly("DATA:WRITE", "DATA:WRITE",
+        "DATA:WRITE");
+    assertThat(authorizedOps.get("data2")).asList().hasSize(3);
+    assertThat(authorizedOps.get("data2")).asList().containsExactly("DATA:WRITE", "DATA:WRITE",
+        "DATA:WRITE");
+
+    Map<String, List<String>> unauthorizedOps = collectVMOps(2, resultsVM0, resultsVM1, resultsVM2);
+
+    assertThat(unauthorizedOps.get("data1")).asList().hasSize(3);
+    assertThat(unauthorizedOps.get("data1")).asList().containsExactly("DATA:WRITE", "DATA:WRITE",
+        "DATA:WRITE");
+  }
+
+  @Test
+  public void clientShouldReAuthenticateWhenCredentialExpiredAndFunctionExecutionOnRegionSucceed()
+      throws Exception {
+    ClientCache clientCache = clientCacheRule.createCache();
+    UpdatableUserAuthInitialize.setUser("data1");
+    Region<Object, Object> region =
+        clientCache.createClientRegionFactory(ClientRegionShortcut.PROXY).create("region");
+    writeFunction = new TestFunctions.WriteFunction();
+
+    ResultCollector rc = onRegion(region).execute(writeFunction);
+    assertThat(((ArrayList) rc.getResult()).get(0))
+        .isEqualTo(TestFunctions.WriteFunction.SUCCESS_OUTPUT);
+
+    // expire the current user
+    VMProvider.invokeInEveryMember(() -> getSecurityManager().addExpiredUser("data1"),
+        serverVM0, serverVM1, serverVM2);
+
+    // do a second function execution, if this is successful, it means new credentials are provided
+    UpdatableUserAuthInitialize.setUser("data2");
+    rc = onRegion(region).execute(writeFunction);
+    assertThat(((ArrayList) rc.getResult()).get(0))
+        .isEqualTo(TestFunctions.WriteFunction.SUCCESS_OUTPUT);
+
+    // all function invocation authorizations are recorded
+    List<Object> resultsVM0 = collectSecurityManagerResults(serverVM0);
```

Comment:
seems like every test needs to combine the results, see `AuthExpirationMultiServerDUnitTest.gatherAuthorizedAndUnauthorizedOps` method, maybe we can make that method as a utility test so that all others can use it.

Reply:
> seems like every test needs to combine the results, see `AuthExpirationMultiServerDUnitTest.gatherAuthorizedAndUnauthorizedOps` method, maybe we can make that method as a utility test so that all others can use it.

Yes, let me do some refactoring to reduce code.